enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+- Operands and CRDs are not automatically removed
+- Upon operator uninstall a user can opt-in to the clean up behavior
+- The operand removal should have a timeout
+- After the operand removal timeout the operator should still be uninstalled
+- Exceeding the operand removal timeout should create an alert event
+
+### Non-Goals
+
+- OLM handles cleanup of anything other than CRs and CRDs
+- Handle custom cleanup logic like filtering/ordering the deletion of CRs 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator user I can instruct OLM to clean up all of my operator's CRs and CRDs when I uninstall my operator.
+
+#### Story 2
+
+As an operator user I want to see an alert event when OLM does not succeed in operand cleanup after a configured timeout.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Specify cleanup via Subscription
+
+To provide a way to opt-in for operand cleanup, the [`Subscription`][subscription] spec can include an optional field `spec.operandCleanup` which lets a user configure the cleanup behaviour.
+
+OLM would initiate cleanup upon the deletion of a `Subscription` instance, based on `spec.operandCleanup`:
+
+##### None (default)
+
+With `spec.operandCleanup.type: none` OLM will not clean up operands. This is also the default behavior if `spec.operandCleanup` is unspecified to make clean up opt-in.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: none
+```
+
+##### OLM managed
+
+With `operandCleanup.type: olm-managed` OLM will delete all CRs, for the operator's owned APIs, in the target namespaces (`olm.targetNamespaces`) of that operator.
+Once the CRs have their finalizers cleared by the operator and are removed from the foreground, OLM will delete the CRD and then proceed with the removal of the operator.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: olm-managed
+    timeout: 30s
+```
+
+Additionaly a timeout configured by `spec.operandCleanup.timeout` would specify how long OLM should wait for the operator to finish cleaning up.
+
+If the cleanup timeout is exceeded, OLM should generate an alert event indicating a cleanup failure for that operator, and then proceed to uninstall the operator.
+
+**Note:** See [Risks and Mitigations section][risks-section] on configuring the timeout.
+
+#### Cleanup for namespace-scoped operators
+
+If there are two non-global operators installed of the same type (providing the same APIs), then removing the CRD during the cleanup of one operator would result in the deletion of CRs for the other operator as well.
+
+To avoid this case OLM would have to ensure that it only removes the CRD during cleanup if:
+
+- The operator installed is global
+- The non-global operator is the only installation or `Subscription` for that operator type in the cluster
+
+#### Operator managed cleanup (not in scope for this proposal)
```

Comment:
I think that operator authors can get this for free if they put finalizers on their CRs, right?

- User deletes subscription
- OLM sends delete to all CRs
- CRs stay around because operator has finalizers on them
- Operator processes them in whatever order / fashion required, and removes the finalizer when done
- CRs actually go away

Reply:
There may be a need for an operator to signal "not ready to be removed" even once all of it's CRs are gone. 

In that case, it could make sense to define a `Deletable` (or `Undeletable`) supported OLM condition according to the [operator condition enhancement](https://github.com/operator-framework/enhancements/blob/master/enhancements/operator-conditions.md)

Reply:
> OLM sends delete to all CRs

What if an operator needs to filter out certain CRs to preserve them from deletion?
The operator can just leave the finalizers intact to keep the CR around but I'm not sure if anything can undo deletion (remove DeletionTimestamp) after OLM sends a blanket delete to all CRs.

Admittedly the use case isn't clear to me but I was thinking of a scenario where the operator has does some pre-processing to decide which CRs to delete.

Although it's a future extension pending on a more concrete use case so I can remove it from this proposal if we think that's okay.

Reply:
I would lean towards the operator signaling `Undeletable` to stop removal, rather than OLM blocking uninstall until it sees a `Deletable` condition as the green light to proceed with operator removal.

It would be a while before most operators would correctly use the operator conditions to signal `Deletable`.

Reply:
Agreed. I think the default case is that an Operator is always `Deletable` unless it is in a very special state where it makes the decision on behalf of the user that deletion of the Operator would cause more pain than denying the deletion of the Operator temporarily. Maybe we should consider a force override on the OLM side for this too, in case Operators are misbehaving and always stay `Undeletable`.